sympy/ntheory/qs.py

```diff
@@ -0,0 +1,489 @@
+
+        Parameters:
+        ===========
+
+        prime : prime number of the factor_base
+        tmem_p : Integer square root of x**2 = n mod prime
+        log_p : Compute Natural Logarithm of the prime
+        """
+        self.prime = prime
+        self.tmem_p = tmem_p
+        self.log_p = log_p
+        self.soln1 = None
+        self.soln2 = None
+        self.a_inv = None
+        self.b_ainv = None
+
+
+def _generate_factor_base(num_primes, n):
+    """Generate factor_base for Quadratic Sieve. The factor_base
+    consists of all the the points whose legendre_symbol(n, p) == 1
+    and p < num_primes. Along with the prime factor_basealso stores
+    natural logarithm of prime and the residue n modulo p.
+
+    Parameters:
+    ===========
+
+    num_primes : upper bound of the factor_base
+    n : integer to be factored
+    """
+    from sympy import prime
+    factor_base = []
+    i = 1
+    while len(factor_base) < num_primes:
+        ith_prime = prime(i)
+        i += 1
+        if pow(n, (ith_prime - 1) // 2, ith_prime) == 1:
+            residue = _sqrt_mod_prime_power(n, ith_prime, 1)[0]
+            log_p = round(log(ith_prime)*2**10)
+            factor_base.append(factor_base_elem(ith_prime, residue, log_p))
+    return factor_base
+
+def _initialize_first_polynomial(N, M, factor_base):
+    """This step is the initialization of the 1st sieve polynomial.
+    Here `a` is selected as a product of several primes of the factor_base
+    such that `a` is about to sqrt(2*N) / M. Other initial values of
+    factor_base elem are also intialized which includes a_inv, b_ainv, soln1,
+    soln2 which are used when the sieve polynomial is changed. The b_ainv
+    is required for fast polynomial change as we don't have to calculate
+    `2*b*mod_inverse(a, prime)` every time.
+
+    Parameters:
+    ===========
+
+    N : Number to be factored
+    M : sieve interval
+    factor_base : factor_base primes
+    """
+    approx_val = sqrt(2*N) / M
+    # `a` is a parameter of the sieve polynomial and `q` is the prime factors of `a`
+    # randomly search for a combination of primes whose multiplication is close to approx_val
+    # This multiplication of primes will be `a` and the primes will be `q`
+    # `best_a` denotes that `a` is close to approx_val in the random search of combination
+    best_a, best_q, best_ratio = None, None, None
+    for _ in range(50):
+        a = 1
+        q = []
+        while(a < approx_val):
+            rand_p = 0
+            while(rand_p == 0 or rand_p in q):
+                rand_p = random.randint(3, len(factor_base) - 1)
+            p = factor_base[rand_p].prime
+            a *= p
+            q.append(rand_p)
+        ratio = a / approx_val
+        if best_ratio is None or abs(ratio - 1) < abs(best_ratio - 1):
+            best_q = q
+            best_a = a
+            best_ratio = ratio
+        # if the best_ratio is < 0.4 it denotes that this `a` chosen is optimal
```

Comment:
There is no mention of this relation in the above ref. link. What is its proof of correctness?

Reply:
While selecting `a`                  i noticed that the ratio        of `a/approx_val` rarely gets close to 0.5 so    i thought if we ever get a best_ratio  < 0.4 then     this `a` can be used. But    this 0.4 is not necessary and can be removed, it shouldn't impact too much on the speed.

Reply:
I don't think we can generalise it without proof or unless you have done some rigorous testing to claim the results (with proofs).

Reply:
cool, then this can be removed.